writeaheadlog_silo_test.go

```diff
@@ -0,0 +1,332 @@
+package writeaheadlog
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"sync"
+	"testing"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/fastrand"
+	"golang.org/x/crypto/blake2b"
+)
+
+type (
+	silo struct {
+		// offset is the file offset in the fake database
+		offset int64
+
+		// numbers
+		numbers []uint32
+
+		// f is the file on which the silo is stored
+		f *os.File
+
+		// i is the index of the number that needs to be incremented next
+		i uint32
+
+		// cs is the checksum of the silo
+		cs checksum
+	}
+
+	siloUpdate struct {
+		// offset is the in the file at which the number should be written
+		offset int64
+
+		// number is the number which is written at index
+		number uint32
+
+		// silo is the offset of the silo so we can apply the update to the right silo
+		silo int64
+	}
+)
+
+// newSilo creates a new silo of a certain length
+func newSilo(offset int64, length int, f *os.File) *silo {
+	if length == 0 {
+		panic("numbers shouldn't be empty")
+	}
+
+	return &silo{
+		offset:  offset,
+		numbers: make([]uint32, length, length),
+		f:       f,
+	}
+}
+
+// marshal marshals a siloUpdate
+func (su siloUpdate) marshal() []byte {
+	data := make([]byte, 24)
+	binary.PutVarint(data[0:8], su.offset)
+	binary.LittleEndian.PutUint32(data[8:12], su.number)
+	binary.PutVarint(data[12:], su.silo)
```

Comment:
best to cast the `int64`s to `uint64` and encoding with `PutUint64` instead of using `Varint`, since `Varint` is a separate variable-length encoding. (Not that it really matters for the purposes of this test, but still.)